### PR TITLE
[Notification] Migrate slack token error email

### PR DESCRIPTION
**Pull request by qnkhuat**
Epic: https://github.com/metabase/metabase/issues/47438

### CODE REVIEW

File: src/metabase/channel/email.clj [hunk 1]
```diff
@@ -122,17 +122,18 @@
...
                                     :notification-recipient/external-email
-                                    [(-> recipient :details :email)]
+                                    [(:email details)]
                                     :notification-recipient/template
-                                    [(-> recipient :details :pattern (channel.params/substitute-params payload))]
+                                    [(not-empty (channel.params/substitute-params (:pattern details) payload (:is_optional details)))]
                                     nil)]
-                     :when (seq emails)]
+                     :when (seq (filter some? emails))]
                  emails)))
```
Comment (crisptrutski) [inline src/metabase/channel/email.clj:137]: I know this doesn't happen in practice yet, but if there are any `nil` emails mixed with non-nil ones, then we'll return the unfiltered list here.

How about rebinding `emails` to be filtered before doing the `when`?

File: src/metabase/channel/params.clj [hunk 2]
```diff
@@ -13,15 +13,18 @@
   "Substitute parameters in text with values from context.
 
   Params are specified using mustache syntax, e.g. {{param}}."
-  [text context]
+ ([text context]
+  (substitute-params text context false))
+ ([text context ignore-missing?]
```
Comment (crisptrutski) [inline src/metabase/channel/params.clj:18]: I'm not sure if there's any practical reason we might add more options in future, but have a naked flag argument is typically a smell and I would prefer kwargs. `(foo! false)` is so much less obvious that `(foo! :fire-missiles false)`.

**Review by crisptrutski — APPROVED** [review]
Agree with Sanya that writing migrations for the templates is going to be a massive pain going forward.

What do you think about having some kind of purely declarative representation, with a generic migration to synchronize?

I'm also curious whether you considered making templates polymorphic to between being in the database or purely defined in code. One advantage of sticking this stuff in the database I guess is that power users can customize it.

**Comment by crisptrutski** [comment]
Fooled by the migration code owners again! Sorry for a premature review.

**Comment by qnkhuat** [comment]
> What do you think about having some kind of purely declarative representation, with a generic migration to synchronize?

let's discuss it here
https://metaboat.slack.com/archives/C07318NQ5M2/p1728273551731969

File: src/metabase/channel/params.clj (hunk 2 above)
Comment (qnkhuat) [reply src/metabase/channel/params.clj]: yep, I do hate naked flag args :D. updated.

File: src/metabase/channel/email.clj (hunk 1 above)
Comment (qnkhuat) [reply src/metabase/channel/email.clj:138]: why not :D updated

**Comment by qnkhuat** [comment]
@crisptrutski it's ready for review again if you want to take another look

**Comment by qnkhuat** [comment]
@crisptrutski  it's ready for review now if you want to take another look

**Review by crisptrutski — APPROVED** [review]
Nice!

**Pull request merged**